Extend diag_mediator to allow the piecemeal posting of diagnostics

ashao opened the pull request:
Some quantities in MOM6 are calculated in subroutines which expect slices of the model's 2d or 3d arrays. Diagnosing these quantities can be challenging because the usual post_data calls expect whole arrays. To solve this problem, the changes here introduce a dynamic buffer that can grow as needed over the course of an simulation.

This buffer keeps track of the diagnostic IDs and the index in the buffer. When a slot in the buffer is no longer needed, for example if the whole array has been computed and posted, the slot is marked as "available" for overwriting. The buffer is only allowed to grow if all the currently allocated slots are in use.

Any computational cost associated with growing this buffer will only happen in the first few steps of the model as post_data for requested diagnostics is called for the first time in that run.

This new piecemeal posting of a diagnostic was implemented for `Kd_ePBL`. The new diagnostic `Kd_ePBL_col_by_col` is the same as the the original field in the `Baltic_OM4_05` case.

@adcroft @marshallward: Two points in particular that I'd like your feedback on:

- The underlying buffer is public, but I cannot think of a way to make it private and still allow a `post_data` call to be done without an extra copy. I have included a `store` subroutine that could be used if we made it private, but otherwise those should be deleted.
- Each time we post the data in a piecemeal fashion, the slot index is computed. We could put in some logic so that we store it once per loop at the beginning of the section where the diagnostic is being computed/posted. I am not sure that it would be worth the effort/complexity for what I am assuming would be a marginal performance gain.

ashao committed:
Create dynamic buffer types for diagnostics posted piecemeal

Some quantities in MOM6 are calculated in subroutines which expect
slices of the model's 2d or 3d arrays. Diagnosing these quantities
can be challenging because the usual post_data calls expect
whole arrays. To solve this problem, the changes here introduce
a dynamic buffer that can grow as needed over the course of an
simulation.

This buffer keeps track of the diagnostic IDs and the index in
the buffer. When a slot in the buffer is no longer needed, for
example if the whole array has been computed and posted, the
slot is marked as "available" for overwriting. The buffer is only
allowed to grow if all the currently allocated slots are in use.

Any computational cost associated with growing this buffer will
only happen in the first few steps of the model as post_data
for requested diagnostics is called for the first time in that
run.

ashao committed:
Implement piecemeal posting of diagnostics

The diag mediator has been extended to add a dynamic buffer to
each axes group. Three new methods have also been added to enable
the piecemeal posting (by column, by point) of a diagnostic and a
'final' method to allow the buffer to be reused later.

ashao committed:
Demonstrate piecemeal diagnostic posting in ePBL

ePBL calculates the vertical diffusivity column by column. This
provides a convenient sanity check of the new piecemeal posting
of diagnostics. The original diagnostic Kd_ePBL is done by posting
the full 3d prognostic array, whereas a new diagnostic
Kd_ePBL_col_by_col posts the same array from within ePBL but does
so column by column.

ashao committed:
Add docstrings for diag_buffer procedures

Fixes failures in the CI due to some procedures and type members
not having docstrings.

Hallberg-NOAA commented on src/framework/MOM_diag_buffers.F90:
```diff
@@ -0,0 +1,421 @@
+  class(diag_buffer_2d), intent(inout) :: this !< This 2d buffer
+
+  integer :: n
+  real, allocatable, dimension(:,:,:) :: temp
```
Please add comments describing the purpose and units (perhaps `[arbitrary]` or '[A ~> a]` of this real variable.

ashao replied:
Comments have been added to indicate that these temporary arrays and the primary `buffer` in the type are arbitrary. The values (and their accordant units) assigned to these buffers will depend on their actual use within the codebase. Whether these represent scaled arbitrary units, depends on how they are filled from within the codebase. As you point out, in the case of their use as part of `post_data`, this may be the case, but may not in general hold.

Hallberg-NOAA commented on src/framework/MOM_diag_buffers.F90:
```diff
@@ -0,0 +1,421 @@
+  integer, dimension(1) :: temp
+  integer :: slot !< The slot in the buffer corresponding to the diagnostic id
+
+  if(allocated(this%ids)) then
```
Please add a space after the `if` here and elsewhere to read `if (...)` to follow the conventions in the MOM6 style guide at https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#white-space.

ashao replied:
Completed

Hallberg-NOAA reviewed (changes requested):
I think that this capability should be extended to include the ability to initialize a diagnostic buffer to land values to avoid the need to call the piecemeal posting for all columns, even over land.

  Also, this PR should be revised to adhere to the guidance in the MOM6 style guide at https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide, as described in specific comments.

ashao committed:
Conform to style for if statements

ashao committed:
Specify that buffer units are arbitrary

ashao committed:
Allow setting of the fill value when expanding the buffer

ashao commented:
@Hallberg-NOAA Thank you for the comments. I have added descriptions of the units (see the inline response) and made the if-statements conform the to the style guide. Additionally, the fill value for the arrays can now be specified; for the purposes of the piecemeal posting of the diagnostics these are set to the `DIAG_MISSING_VALUE` specified in `MOM_input`.

Hallberg-NOAA commented on src/framework/MOM_diag_buffers.F90:
```diff
@@ -0,0 +1,432 @@
+  integer :: js !< The start slot of the array j-direction
+  integer :: ie !< The end slot of the array i-direction
+  integer :: je !< The end slot of the array j-direction
+  real :: fill_value = 0. !< Set the fill value to use when growing the buffer
```
There need to be units ("[arbitrary]") in the comment describing `diag_buffer_base%fill_value`.

ashao committed:
Refactor buffers based on Marshall's suggestions

Hallberg-NOAA commented on src/framework/MOM_diag_buffers.F90:
```diff
@@ -0,0 +1,472 @@
+!> Set the fill value to use when growing the buffer
+subroutine set_fill_value(this, fill_value)
+  class(diag_buffer_base), intent(inout) :: this !< The diagnostic buffer
+  real,                    intent(in)    :: fill_value !< The fill value to use when growing the buffer [arbitrary]
```
I believe that this subroutine is missing the line `this%fill_value = fill_value`, without which it does not do anything.

ashao replied:
Thanks for pointing this out (line was lost in a refactor). I added a test to make sure that this actually does what we expect

ashao committed:
Actually set the fill value and add tests

Hallberg-NOAA reviewed (approved):
With this latest set of changes, all of my concerns have been addressed, and I think that this PR is now ready to be merged in.

marshallward commented on src/framework/MOM_diag_buffers.F90:
```diff
@@ -0,0 +1,513 @@
+
+    allocate(test_2d(nlen, is:ie, js:je))
+    call random_number(test_2d)
+    buffer = diag_buffer_2d(is=is, ie=ie, js=js, je=je)
```
This is causing an error in the Nvfortran compiler:
```
NVFORTRAN-F-0155-No default initialization in structure constructor- member buffer (../../../../.././/src/MOM6/src/framework/MOM_diag_buffers.F90: 326)
```
This is happening in nvhpc 23.9.

Not sure exactly what is causing the error, but the structure constructor may be unsupported.  Can you explicitly assign these `is`/`ie`/`js`/`je` values?

Hallberg-NOAA reviewed (changes requested):
Please modify the syntax of this code so that it compiles and works correctly with all 3 of the intel, gnu and nvidia compilers.

ashao committed:
Fix issue with partial construction of buffer

Hallberg-NOAA commented:
Unfortunately we are now getting segmentation faults with all three of the intel, gnu and nvidia compilers when we try to run the pipeline testing with this PR.  The GFDL regression testing at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/jobs/163289 seems to indicate that the problem in various test cases arises with the `allocate()` statement in `grow_3d()` on line 225 of MOM_diag_buffers.F90, or perhaps with the `allocated(this%ids)` test on line 105 in `find_buffer_slot()`, or on the reference to `this%ids(:)` on line 125 in `grow_ids()`.  I suspect that these are arising from calls with an unallocated `diag_buffer` object.  However this did pass our TC testing, so I will try this again later in case these problems are arising because of a temporary problem with the Gaea computer.  In the meantime, @ashao, if you get the chance to revisit this, that would be appreciated.

Hallberg-NOAA commented:
I went back and looked at the original regression testing that we did back in late February at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/26532.  That version did not compile with the Nvidia compiler, but it did compile with gnu and intel.  When it ran, it was also getting segmentation faults and bus errors. We focused on the compile failures, but the run-time errors were also problematic, and we should have pointed these out at the same time.  The failed tests include everything using ePBL, the simplest of which would be something like ocean_only/single_column/EPBL.

ashao committed:
Fixes logic error when using piecemeal buffers

Two issues were leading to sporadic compilation errors and
segmentation faults

1) The post_diagnostics_by_point only looked up the slot in the
buffer but would not expand the buffer -> now uses
check_capacity_by_id which will expand the buffer if needed

2) post_data_3d_final now conditionally checks to make sure that
the buffer slot is not zero before taking any action

ashao commented:
Clsoing while debugging

ashao committed:
Fix missing check on non-negative diagnostic ID

Fixes a bug where the Kd_ePBL_col_by_col diagnostic's post_data
calls were not being checked before being called.

Also, move setting of extents for buffers on native axes to mask
generation

ashao committed:
Initialize extents after set_axes_info

Update the extents for the piecemeal buffers after the axes info
is set. Resolves a problem where the extents were not being
retained.

ashao commented:
@Hallberg-NOAA @marshallward: These changes should resolve the segmentation faults. Additionally, I have confirmed that the `Kd_ePBL_col_by_col` and the `Kd_ePBL` diagnostics are the same (as long as the the fill value is 0.).

Hallberg-NOAA commented on src/framework/MOM_diag_buffers.F90:
```diff
@@ -0,0 +1,551 @@
+      local_fail = local_fail .or. (ubound(buffer%buffer(i)%field, 2) /= je)
+      local_fail = local_fail .or. (lbound(buffer%buffer(i)%field, 3) /= ks)
+      local_fail = local_fail .or. (ubound(buffer%buffer(i)%field, 3) /= ke)
+    if (verbose) write(stdout,*) "grow_buffer_3d: ", local_fail
```
It looks to me like this line is duplicative with the message being written out two lines below (on line 455).

Hallberg-NOAA replied:
We can remove the duplicative informational messages about grow_buffer_3d later if they become too disruptive.

Hallberg-NOAA reviewed (approved):
This PR is now passing all of our testing. The pipeline testing has passed at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/30328 with the expected warning about the addition of a new diagnostic.
